Safe method to get the repr of an object.

Carreau opened the pull request:
- Protect against recursion.
 - <del>Incorrect because of bugs in IPython, in particular the priority of
   formatters is not well respected.</del>

First step toward #10647

Carreau committed:
[WIP] Safe method to get the repr of an object.

 - Protect against recursion.
 - Incorrect because of bugs in IPython, in particular the priority of
   formatters is not well respected.

First step toward #10647

takluyver commented on IPython/core/display.py:
```diff
@@ -140,6 +141,88 @@ def _new_id():
+
+
+
+def get_repr(obj, include=None, exclude=None, *, on_recursion=_RecursiveObject):
```
Naming: should this be something like `get_repr_mimebundle()` or `get_rich_reprs()`? It seems like a good idea to clearly distinguish our rich repr system from Python's text-based reprs.

takluyver commented on IPython/core/display.py:
```diff
@@ -140,6 +141,88 @@ def _new_id():
+
+    Returns
+    -------
+    (format_dict, metadata_dict) : named tuple of two dicts
```
I think the names here should match what they are in the named tuple, i.e. `(data, metadata)`, rather than involving a different pair of names for them.

takluyver commented on IPython/core/display.py:
```diff
@@ -140,6 +141,88 @@ def _new_id():
+
+        0/data: format_dict is a dictionary of key/value pairs, one of each
+        format that was generated for the object. The keys are the format
+        types, which will usually be MIME type strings and the values and
```
> usually be MIME type strings

Are there circumstances in which they're not? If so, we should make at least a brief mention of what they are. If we just mean that they might be unofficial mime types, I think that's OK - unregistered mime types beginning with either `x-` or `vnd.` are common enough that they don't need special mention.

Should also mention:

- For most mime types (other than `application/*json`), values will be strings.
- Binary data will be base64 encoded strings (assuming this handles that step), and it's up to calling code to know which are binary. [aside: should we standardise a field in metadata to indicate which mime types are base64 encoded?]

Carreau replied:
that's copy/pasted from the formatter module, I'm guessing we should either synchronize, of get the information in a single place.

takluyver commented on IPython/core/display.py:
```diff
@@ -140,6 +141,88 @@ def _new_id():
+    the representation of `_RecursiveObject`
+
+    """
+    global _objs
```
I'm not keen on the use of a global store here - you could get weird results if there's more than one thread or coroutine using this function at once. I'll try to think of a better way to handle the recursion thing.

Carreau replied:
I don't see any way that does not require a global-ish state, I can move it into a class (that would be better).
But we can't really enforce users to pass variable down. 

I'm still expecting this to almost never go on and recurse. 
If you are concern about threading I can use  `threading.local()`. Coroutines: there is no tasklocals in Python (yet?) and having a task-switch would require having coroutiness based reprs, which I'm unsure we do support. 

Worse case is I believe we will go down the recursion path where there is no recursions, (AFAICAT), and I think it is a reasonable tradeoff.

Carreau commented:
Refactored, cleaned and documented with some tests. Still some polish to do. 
I moved the "get_repr_mimebundle" into a class method with "use your own instance of the class if you have any chance of having concurrency", Of course you can shoot your self in the foot, by creating a instance everytime you compute the repr, but he, you can shoot yourself in the foot in so many different way, and you can also call the formatters yourself.

I now also "Fake" the `__module__` object of everything in IPython.core.display as otherwise Sphinx can't cross link (I'm partially unhappy because it changes the `repr`.)

The DataMetadata now have `.data` and `.metadata` attributes instead of being a named tuple.

The `IPython.display` module have some top-level docs.

takluyver commented on IPython/core/display.py:
```diff
@@ -139,6 +141,117 @@ def _new_id():
+
+    @property
+    def data(self):
+        return self[0]
```
Sorry, I didn't follow why this is preferable to a NamedTuple?

Carreau replied:
I had issue with namedtuple, where it keep saying in soma place the `.data` was not defined but `[0]` was working. I'm confused as to exactly why but I can retry.

takluyver commented on IPython/core/display.py:
```diff
@@ -1338,3 +1448,15 @@ def set_matplotlib_close(close=True):
+    # that's a module we explicitly do not document.
+    try:
+        # using hasattr to check still raises an AttributeError.
+        locals()[oname].__module__ = 'IPython.display' 
```
Eugh. Maybe we can figure out improvements to Sphinx...

Carreau replied:
Or, we re-allow `IPython.core.display` to be documented. Right now sphinx can't link because there is no `IPython.core.display` page.

takluyver replied:
What happens if we cross reference to e.g. 

    :func:`IPython.display.display`

minrk replied:
Functions defined in one module can be referenced in any module in which they are importable, we just have to make sure that they are present in the autodoc module for the desired target.

I do this [in pyzmq](https://github.com/zeromq/pyzmq/blob/master/docs/source/api/zmq.rst), where docs treat everything as being in top-level `zmq` module, but classes are actually defined all over the place.

Carreau commented:
Not sure, but that won't help with the base class that are automatically
inferred.

On Jun 13, 2017 08:14, "Thomas Kluyver" <notifications@github.com> wrote:

> *@takluyver* commented on this pull request.
> ------------------------------
>
> In IPython/core/display.py
> <https://github.com/ipython/ipython/pull/10651#discussion_r121708575>:
>
> > @@ -1338,3 +1448,15 @@ def set_matplotlib_close(close=True):
>      from ipykernel.pylab.config import InlineBackend
>      cfg = InlineBackend.instance()
>      cfg.close_figures = close
> +
> +
> +
> +for oname in __all__:
> +    # set the module name to IPython.display or sphinx cannot correctly cross
> +    # link. Mostly it figures out that objets are in IPython.core.display, but
> +    # that's a module we explicitly do not document.
> +    try:
> +        # using hasattr to check still raises an AttributeError.
> +        locals()[oname].__module__ = 'IPython.display'
>
> What happens if we cross reference to e.g.
>
> :func:`IPython.display.display`
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ipython/ipython/pull/10651#discussion_r121708575>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAUezw-HNO-gOcZhrh8zvqSNXDsEgdUZks5sDqdBgaJpZM4N11Yr>
> .
>

Carreau commented:
TODO: Look at wether a custom formatter can say "I actually don't know how to represent this".

Carreau commented:
meh, I'm un-happy with the actual API when using it. I keep using `get_repr_mimebundle(obj).data.get('text/html', None) or escape(repr(obj))` to build my own nested repr. I defined a helper function, but that seem less than optimal. Though it allows to do some nice things. I'm going to think a bit about the API.

minrk commented on IPython/core/display.py:
```diff
@@ -139,6 +141,117 @@ def _new_id():
+# Expose this for convenience at the top level. Similar to what the random
+# module in python does. If you want to avoid weird behavior from concurrency:
+#   Instantiate your own.
+get_repr_mimebundle = ReprGetter().get_repr_mimebundle
```
If `get_repr_mimebundle` instantiated an internal ReprGetter at call time rather than being a method on a global object, there shouldn't be a potential for race conditions.

Carreau replied:
but that would kill the detection of recursion, as every call would create a new repr_getter that would track independently which objects have been seen.

minrk replied:
It shouldn't. The recursion would be tracked for the duration of the recursion, and the instance disposed of when the outermost call finishes.

Carreau replied:
I'm confused. 


Like so ?
```python
def get_repr_mimebundle(*args, **kwargs):
      return ReprGetter.instance().get_repr_mimebundle(*args, **kwargs)
```
?

I don't see how that prevent race condition. 

If it unconditionally create an instance, and an object have:

```python
from IPython.display import get_repr_mimebundle
class MyContainer:

    def _repr_html_(self):
          return '|'.join(get_repr_mimebundle(obj)['text/html'] for obj in self.objects)
```
Then in breaks recursion detection, and is not thread/task safe as you get the global get_repr_mimebundle anyway. 

Or I'm missing something.

minrk replied:
Another option is to use threadlocal storage for the global. Then there is no way to have a race condition because all of the cooperative concurrency libraries (asyncio, etc.) can't have a race because this is a blocking call.

minrk commented on IPython/core/display.py:
```diff
@@ -27,7 +27,8 @@
 'Image', 'clear_output', 'set_matplotlib_formats', 'set_matplotlib_close',
-'publish_display_data', 'update_display', 'DisplayHandle', 'Video']
+'publish_display_data', 'update_display', 'DisplayHandle', 'Video', 'get_repr_mimebundle',
+'RecursiveObject', 'ReprGetter']
```
RecursiveObject and ReprGetter should probably *not* be added to `__all__`, since they are private, internal implementation details.

Carreau replied:
Fair enough for it to be clean, but I still expect people to need to register formatter for `RecursiveObject` if they want nice representation for format it does not support currently, so you need a handle on it.  ReprGetter also need to be public if you are computing recursive Repr on multiple task/threads

Carreau replied:
removed.

Carreau commented on IPython/core/formatters.py:
```diff
@@ -139,6 +139,9 @@ def format(self, obj, include=None, exclude=None):
             precedence and the corresponding `_repr_*_` for this mimetype will
             not be called.
 
+            TODO: Figure out and document whether user registered formatters do
```
Yes, in theory it seem to not be the case with `_repr_mimebundle_`, I should have been clearer in the comment.

Carreau commented:
Note to self, the recursion stuff also does not work if a formatter need another. Like for example, the HTML, being "just" a bold version of the text/plain.

Carreau commented:
> Note to self, the recursion stuff also does not work if a formatter need another. Like for example, the HTML, being "just" a bold version of the text/plain.

Ok, I got that to work.

Carreau committed:
Track recursion on a per-mimetype basis.

This allow richer formatters to requires reprs from less-rich
formatters

Carreau committed:
Check parameter of `for_type_by_name`

It is easy to assume the first parameter is a fully-qualified name.

minrk commented on IPython/core/display.py:
```diff
@@ -139,6 +142,124 @@ def _new_id():
+
+
+
+DataMetadata = namedtuple('DataMetadata', ('data','metadata'))
```
You were asking about the issue of conveniently retrieving mime-keys with the `('data', 'metadata')` tuple. One option is a method on this object:

```python
def get_mime(self, mimetype):
    """Extract a single mime-type representation from a mimebundle"""
    if mimetype not in self.data:
        return (None, None)
    else:
        return self.data[mimetype], self.metadata.get(mimetype)
```

This could also be wrapped in a single function:

```python
def get_repr_mime(obj, mimetype):
    """Get the representation of an object as a single mime-type"""
    return get_repr_mimebundle(obj).get_mime(mimetype)
```

A `metadata=False` flag could also prevent it from returning the metadata part, if you want to avoid getting the first item for the data part.
